### PR TITLE
Enable Plan Upgrades

### PR DESCRIPTION
**Pull request by glrf**
## Checklist

- [x] Keep pull requests small so they can be easily reviewed.
- [x] Categorize the PR by setting a good title and adding one of the labels:
  `bug`, `enhancement`, `documentation`, `change`, `breaking`,
  as they show up in the changelog
- [x] Update the documentation.
- [x] Update tests.

### CODE REVIEW

File: pkg/brokerapi/broker.go [hunk 1]
```diff
@@ -332,34 +334,21 @@ func (b Broker) Update(rctx *reqcontext.ReqContext, instanceID, serviceID, oldPl
...
-		if instanceSLA == crossplane.SLAPremium && newPlanSLA == crossplane.SLAStandard {
-			return true
-		}
-		if instanceSLA == crossplane.SLAStandard && newPlanSLA == crossplane.SLAPremium {
-			return true
-		}
-		return false
+	cmp, err := p.Cmp(*np)
+	if err != nil {
+		return res, ErrPlanChangeNotPermitted
```
Comment (simu) [inline pkg/brokerapi/broker.go:339]: Maybe add a clarifying comment here to note which conditions this return covers.

```suggestion
		// The new plan has a different service, unknown service, or unknown size.
		// This change is not permitted.
		return res, ErrPlanChangeNotPermitted
```

**Review by simu — APPROVED** [review]
LGTM overall. I did not read the changes in `pkg/brokerapi/brokerapi_test.go` in detail.

File: pkg/brokerapi/broker.go [hunk 2]
```diff
@@ -332,34 +334,21 @@ func (b Broker) Update(rctx *reqcontext.ReqContext, instanceID, serviceID, oldPl
...
+		// The new plan is smaller than the old.
+		// This is not supported.
+		return res, ErrPlanChangeNotPermitted
+	}
+	if cmp < 0 && !b.allowPlanUpgrades {
+		// The new plan is large than the old.
+		// This is only supported if explicitly enabled.
 		return res, ErrPlanChangeNotPermitted
 	}
+	// Plans only differ in SLA
```
[the remark this answers is not in the document]
Comment (glrf) [reply pkg/brokerapi/broker.go:351]: That's true. I expect someone to think about this if they introduce a new SLA, but maybe a comment here would be helpful.

What I just realized it, that the old implementation returned an error, if someone tries to switch a service to the same plan it already has. I'm not sure if this was intended, as it is not tested. I personally think returning success is a better API design, but that would strictly speaking be a breaking change?

[the remark this answers is not in the document]
Comment (simu) [reply pkg/brokerapi/broker.go:353]: Might be worth to check if anybody is relying on the error when trying to change an instance to the same plan it already has.

If not, we don't really need to treat allowing "no-op" changes as a breaking change.

**Comment by glrf** [comment]
With the new requirements I needed to completely restructure the upgrade validity check. I might have gone a little overboard with how generic the solution is. 

We can now allow any subset of plan size changes with a white-list approach:

```
OSB_PLAN_UPDATE_SIZE_RULES="xsmall>small|medium>large"
```

This would only allow plan changes from `xsmall` to `small` and from `medium` to `large`.  I will definitely document this thoroughly, but maybe its generally a bad idea, too generic, and too confusing?

**Review by simu — APPROVED** [review]
I like the new approach. I don't think it's too complex or confusing. If it's well-documented, I think it's a very good way to make the set of allowed changes configurable.

**Pull request merged**